pkg/dockerregistry/client.go

```diff
@@ -597,6 +603,87 @@ func (repo *v2repository) getImage(c *connection, image, userTag string) (*Image
 	return repo.getTaggedImage(c, image, userTag)
 }
 
+func (repo *v2repository) getImageConfig(c *connection, dgst string) ([]byte, error) {
+	endpoint := repo.endpoint
+	endpoint.Path = path.Join(endpoint.Path, fmt.Sprintf("/v2/%s/blobs/%s", repo.name, dgst))
+	req, err := http.NewRequest("GET", endpoint.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+
+	if len(repo.token) > 0 {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", repo.token))
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, convertConnectionError(c.url.String(), fmt.Errorf("error getting image config for %s: %v", repo.name, err))
+	}
+	defer resp.Body.Close()
+
+	switch code := resp.StatusCode; {
+	case code == http.StatusUnauthorized:
+		if len(repo.token) != 0 {
+			// The DockerHub returns JWT tokens that take effect at "now" at second resolution, which means clients can
+			// be rejected when requests are made near the time boundary.
+			if repo.retries > 0 {
+				repo.retries--
+				time.Sleep(time.Second / 2)
+				return repo.getImageConfig(c, dgst)
+			}
+			delete(c.cached, repo.name)
+			// docker will not return a NotFound on any repository URL - for backwards compatibility, return NotFound on the
+			// repo
+			body, _ := ioutil.ReadAll(resp.Body)
+			glog.V(4).Infof("passed valid auth token, but unable to find image config at %q, %d %v: %s", req.URL.String(), resp.StatusCode, resp.Header, body)
+			return nil, errBlobNotFound{dgst, repo.name}
+		}
+		token, err := c.authenticateV2(resp.Header.Get("WWW-Authenticate"))
+		if err != nil {
+			return nil, fmt.Errorf("error getting image config for %s:%s: %v", repo.name, dgst, err)
+		}
+		repo.retries = 2
+		repo.token = token
+		return repo.getImageConfig(c, dgst)
+	case code == http.StatusNotFound:
+		body, _ := ioutil.ReadAll(resp.Body)
+		glog.V(4).Infof("unable to find image config at %q, %d %v: %s", req.URL.String(), resp.StatusCode, resp.Header, body)
+		return nil, errBlobNotFound{dgst, repo.name}
+	case code >= 300 || resp.StatusCode < 200:
+		// token might have expired - evict repo from cache so we can get a new one on retry
+		delete(c.cached, repo.name)
+
+		return nil, fmt.Errorf("error retrieving image config: server returned %d", resp.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("can't read image body from %s: %v", req.URL, err)
+	}
+
+	return body, nil
```

Comment:
What worries me the most is this whole code block is a copy&pase of `getTaggedImage`. I'm ok with this being in but we need to merge the two into single method.